---
title: "BIO477 R workshop"
author: "Katherine I. Cooper"
date: "2025-11-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Tutorial to create differential gene expression csv and visuals

This is a guided view of the differential gene expression pipeline by [Bioconductor](https://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf).

No data is required for this tutorial. If implementing into other research, an RNA-seq feature counts file is required.

------------------------------------------------------------------------

#### Load packages and data

Load required libraries. If an error code shows up saying the library is "Not available for this version of R," the package likely has to be installed using BiocManager::install("*package_name_here"*).

```{r}
library(edgeR)

```

**Preparing the data**

We are using the recount3 package, a reservoir of public SRA data. The data comes in a SummarizedExperiment file and needs to be prepared prior to using the pipeline.

```{r}
breast_data<-recount3::create_rse_manual(
  project = "SRP042620",
  project_home = "data_sources/sra",
  organism = "human",
  annotation = "gencode_v26",
  type = "gene") #Publicly available triple negative breast cancer reads. Paired with other tumor types as well, including immoratalized cell-lines, ER+ tumors, and tissue adjacent to the tumors.


assay(breast_data, "counts") <- transform_counts(breast_data)

feature_counts <- assay(breast_data, "counts")
rownames(feature_counts) <- sub("\\.\\d+$", "", rownames(feature_counts)) #Cleaning the names.

group <- colData(breast_data)$sra.sample_attributes %>%
  sub("^source_name;;([^|]+).*", "\\1", .) %>%
  factor()
#Pulling the condition names from the SummarizedExperiment file and formatting the group names to be more readable.
```

```{r}
View(group) #View the groups present in the data.
```

------------------------------------------------------------------------

#### Manipulate data

*Group:* Creates the treatment assignments. Insert your treatments in the *c(...)* part of the line and the number of replicates after the *each* part. Run the group line under to double check your treatments are as they should be.

The variable *y* is where the DGEList will be stored. This has the data of the differential genes. The *counts* clause is where your feature counts (fc) should be, and the *group* clause is your treatment groups.

```{r}
 
y <- DGEList(counts = feature_counts, group = group)
# colnames(y) #SRA numbers corresponding to the sequence read.
```

The *colnames(y)* will return your column names from the original feature counts. To view the architecture of 'y', type head(y) into your console. The samples table will have your treatment groups. To view, remove the \# and run the line. The column names represent the different sequences which match with the treatment names.

Now that we have the DGEList, we want to rename the genes from Ensembl IDs, the form they are imported as, to symbol IDs. For example, ENSG00000139618 is the Ensembl ID corresponding to BRCA2.

```{r}
require(org.Hs.eg.db) #Contains an annotated human genome. Translates gene names to more common names.
Symbol <- mapIds(org.Hs.eg.db, keys=rownames(y), keytype="ENSEMBL", column = "SYMBOL")
y$genes <- data.frame(Symbol=Symbol)

#Check the names.
y$genes
```

------------------------------------------------------------------------

#### Filtering and normalization

Filtering and normalization steps are important to ensure confounding variables don't skew results.

*filterByExpr(y)* removes genes with negligble expression. *normLibSizes(y)* adjusts the samples to make up for library size differences. This is to not impact the logFoldChange later on.

```{r}
#Filtering and normalization
keep <- filterByExpr(y)
summary(keep)

y <- normLibSizes(y)

plotMD(cpm(y, log=TRUE), column=1) #Change column number to see how different samples perform. 
abline(h=0, col="red", lty=2, lwd=2)
```

The MD plot shows the quality of the reads. Most genes should be centered around the red line. To ensure quality of all samples, cycle through them using the *column = x* clause. The data set we are using already has high quality reads thanks to it being published and accumulated in the reecount3 package.

------------------------------------------------------------------------

#### Choose contrasts

Prior to choosing contrasts, a design matrix needs to be created. The levels may be out of order. That is okay. If you want to double check the labels prior to adding column names, run the design command before the *colnames(design)* command. The common dispersion shows the "squared coefficient of variation."

```{r}
#Design matrix
design <- model.matrix(~0 + group)
colnames(design) <- levels(group) #If factor() was not used when making the groups, an error code will appear here. 

#Dispersion estimation
y <- estimateDisp(y, design, robust=TRUE) #This may take a few minutes.
y$common.dispersion
plotBCV(y)
fit <- glmQLFit(y, design, robust=TRUE)
fit$dispersion
plotQLDisp(fit)
```

This is where you chose your contrast levels. Here, we are comparing the cancer-free condition to the triple negative breast cancer (TNBC) primary tumor. All the positive genes resulting from this have higher expression in the cancer-free condition.

Contrasts can be changed to fit the needs of your project. Replace the current contrasts with any of the following.

Copy and paste into the code to explore the DGE.

-   Breast.Cancer.Cell.Line
-   ER..Breast.Cancer.Primary.Tumor
-   Reduction.Mammoplasty...No.known.cancer
-   Triple.Negative.Breast.Cancer.Primary.Tumor
-   Uninvolved.Breast.Tissue.Adjacent.to.ER..Primary.Tumor
-   Uninvolved.Breast.Tissue.Adjacent.to.TNBC.Primary.Tumor

These correspond to the column names in the design. 1) Immortalized cell lines used in breast cancer research, 2) ER+ primary tumor, 3) Cancer-free sample, 4) TNBC primary tumor, 5) Tissue adjacent to the ER+ primary tumor, 6) Tissue adjacent to TNBC primary tumor.

```{r}
colnames(design) <- make.names(colnames(design)) #Converts the spaces in the column names into "."
con <- makeContrasts(Reduction.Mammoplasty...No.known.cancer - Triple.Negative.Breast.Cancer.Primary.Tumor, levels=design) #This is where you specify the contrasts you want to use.
qlf <- glmQLFTest(fit, contrast = con)


DGE_TNBC <- topTags(qlf, n =Inf)
head(DGE_TNBC) #Differential genes
```

The *topTags()* function identifies the top deferentially expressed genes. The pipeline for DGE is now complete. Below is the code to save the DGE list to a CSV file to store for later analyses. Because we are transitioning immediately, this step can be omitted.

```{r}
#To save to a csv file
write.csv(DGE_TNBC, "~/TNBC_DGE.csv", col.names = TRUE)

```

# Gene ontology pathway

Moving onto gene ontology, this is where our now identified differential genes will be grouped into their respective classification. We will be using the DGE_TNBC file we created above.

```{r}
#Install libraries 
BiocManager::install("clusterProfiler")
BiocManager::install("pathview")
BiocManager::install("enrichplot")
library(clusterProfiler)
library(enrichplot)
library(ggplot2)



DGE_gene_list <- DGE_TNBC$table 
geneList <- DGE_gene_list$logFC
names(geneList) <- rownames(DGE_gene_list)
geneList <- sort(geneList, decreasing = TRUE)
geneList <- na.omit(geneList)


gse <- gseGO(geneList=geneList, 
             ont ="ALL", 
             keyType = "ENSEMBL", 
             nPerm = 10000, 
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = org.Hs.eg.db, 
             pAdjustMethod = "none")
require(DOSE)
dotplot(gse, showCategory=5, split=".sign") + facet_grid(.~.sign) 


```
