---
title: "Ontology R workshop"
author: "Katherine I. Cooper"
date: "2025-11-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE)
```

# Tutorial to create differential gene expression csv and visuals

This is a guided view of the differential gene expression pipeline by [Bioconductor](https://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf).

No data is required for this tutorial. If implementing into other research, an RNA-seq feature counts file is required.

------------------------------------------------------------------------

### Load packages and data

**Windows users need a package called Rtools.** Please download [here](https://cran.r-project.org/bin/windows/Rtools/)

Load required libraries. If an error code shows up saying the library is "Not available for this version of R," the package likely has to be installed using BiocManager::install("*package_name_here"*).

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

# CRAN packages
Base_packages <- c("ggplot2", "ggrepel", "tidyverse")

for (pkg in Base_packages) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    install.packages(pkg)
  }
  library(pkg, character.only = TRUE)
}

# Bioconductor packages
Bioconductor_packages <- c(
  "edgeR", "recount3", "clusterProfiler",
  "pathview", "enrichplot", "org.Hs.eg.db", "GenomeInfoDb"
)

for (pkg in Bioconductor_packages) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    BiocManager::install(pkg, update = TRUE, ask = FALSE)
  }
  library(pkg, character.only = TRUE)
}

# List loaded packages
all_pkgs <- c(Base_packages, Bioconductor_packages)
cat("The following packages are now loaded:\n")
cat(paste0(" - ", all_pkgs, collapse = "\n"))

```

Prior to moving on, please **ensure all packages** are loaded:

-   Base packages: ggplot2, ggrepel, tidyverse

-   Bioconductor packages: edgeR, recount3, clusterProfiler, pathview, enrichplot, org.Hs.eg.db

There are **10** total.

If they are missing, type in the console: *install.packages("InsertPackageName")* or *BiocManager::install("InsertPackageName")*

Then, load the library using *library("PackageName").*

*Alternatively, check this Google Doc for the code needed*: <https://docs.google.com/document/d/1vl-nIzDotmrF0a8wrI9N7dUBzDXg5_-UIh_72EdCoGw/edit?usp=sharing>

### **Preparing the data**

We are using the recount3 package, a reservoir of public SRA data. The data comes in a SummarizedExperiment file and needs to be prepared prior to using the pipeline.

**A prompt will appear asking you to create a directory; say yes.**

```{r}
breast_data<-recount3::create_rse_manual(
  project = "SRP042620",
  project_home = "data_sources/sra",
  organism = "human",
  annotation = "gencode_v26",
  type = "gene") 
#Publicly available triple negative breast cancer reads. Paired with other tumor types as well, including immoratalized cell-lines, ER+ tumors, and tissue adjacent to the tumors.


assay(breast_data, "counts") <- transform_counts(breast_data)

feature_counts <- assay(breast_data, "counts")
rownames(feature_counts) <- sub("\\.\\d+$", "", rownames(feature_counts)) #Cleaning the names.

group <- colData(breast_data)$sra.sample_attributes %>%
  sub("^source_name;;([^|]+).*", "\\1", .) %>%
  factor()
#Pulling the condition names from the SummarizedExperiment file and formatting the group names to be more readable.
```

------------------------------------------------------------------------

### Manipulate data

*Group:* Creates the treatment assignments. Insert your treatments in the *c(...)* part of the line and the number of replicates after the *each* part. Run the group line under to double check your treatments are as they should be.

The variable *y* is where the DGEList will be stored. This has the data of the differential genes. The *counts* clause is where your feature counts (fc) should be, and the *group* clause is your treatment groups.

```{r}
 
y <- DGEList(counts = feature_counts, group = group)
# colnames(y) #SRA numbers corresponding to the sequence read.
```

The *colnames(y)* will return your column names from the original feature counts. To view the architecture of 'y', type head(y) into your console. The samples table will have your treatment groups. To view, remove the \# and run the line. The column names represent the different sequences which match with the treatment names.

Now that we have the DGEList, we want to rename the genes from Ensembl IDs, the form they are imported as, to symbol IDs. For example, ENSG00000139618 is the Ensembl ID corresponding to BRCA2.

When you check the names, there will be some N/A values. This is okay.

```{r}
require(org.Hs.eg.db) #Contains an annotated human genome. Translates gene names to more common names.
Symbol <- mapIds(org.Hs.eg.db, keys=rownames(y), keytype="ENSEMBL", column = "SYMBOL")
y$genes <- data.frame(Symbol=Symbol)

#Check the names.
y$genes
```

------------------------------------------------------------------------

### Filtering and normalization

Filtering and normalization steps are important to ensure confounding variables don't skew results.

*filterByExpr(y)* removes genes with negligble expression. *normLibSizes(y)* adjusts the samples to make up for library size differences. This is to not impact the logFoldChange later on.

```{r}
#Filtering and normalization
keep <- filterByExpr(y)
summary(keep)

y <- normLibSizes(y)

plotMD(cpm(y, log=TRUE), column=1) + #Change column number to see how different samples perform. 
  abline(h=0, col="red", lty=2, lwd=2)
```

The MD plot shows the quality of the reads. Most genes should be centered around the red line. To ensure quality of all samples, cycle through them using the *column = x* clause. The data set we are using already has high quality reads thanks to it being published and accumulated in the reecount3 package.

------------------------------------------------------------------------

### Choose contrasts

Prior to choosing contrasts, a design matrix needs to be created. The levels may be out of order. That is okay. If you want to double check the labels prior to adding column names, run the design command before the *colnames(design)* command. The common dispersion shows the "squared coefficient of variation."

```{r}
#Design matrix
design <- model.matrix(~0 + group)
colnames(design) <- levels(group) #If factor() was not used when making the groups, an error code will appear here. 

#Dispersion estimation
y <- estimateDisp(y, design, robust=TRUE) #This may take a few minutes.
y$common.dispersion
plotBCV(y)
fit <- glmQLFit(y, design, robust=TRUE)
fit$dispersion
plotQLDisp(fit)
```

This is where you chose your contrast levels. Here, we are comparing the triple negative breast cancer cells (TNBC) to the healthy, cancer-free cells. All the positive genes are upregulated in the cancer environment.

Contrasts can be changed to fit the needs of your project. Replace the current contrasts with any of the following.

Copy and paste into the code to explore the DGE.

-   Breast.Cancer.Cell.Line
-   ER..Breast.Cancer.Primary.Tumor
-   Reduction.Mammoplasty...No.known.cancer
-   Triple.Negative.Breast.Cancer.Primary.Tumor
-   Uninvolved.Breast.Tissue.Adjacent.to.ER..Primary.Tumor
-   Uninvolved.Breast.Tissue.Adjacent.to.TNBC.Primary.Tumor

These correspond to the column names in the design. 1) Immortalized cell lines used in breast cancer research, 2) ER+ primary tumor, 3) Cancer-free sample, 4) TNBC primary tumor, 5) Tissue adjacent to the ER+ primary tumor, 6) Tissue adjacent to TNBC primary tumor.

```{r}
colnames(design) <- make.names(colnames(design)) #Converts the spaces in the column names into "."
con <- makeContrasts(Triple.Negative.Breast.Cancer.Primary.Tumor - Reduction.Mammoplasty...No.known.cancer, levels=design) #This is where you specify the contrasts you want to use.
qlf <- glmQLFTest(fit, contrast = con)


DGE_TNBC <- topTags(qlf, n =Inf)
head(DGE_TNBC) #Differential genes
```

The *topTags()* function identifies the top deferentially expressed genes. The pipeline for DGE is now complete. Below is the code to save the DGE list to a CSV file to store for later analyses. Because we are transitioning immediately, this step can be omitted.To save the file, use "write.csv(DGE_TNBC, insert_pathname_here, row.names = TRUE)"

### Plotting volcano plot

First, we need to sort our the table from the other information stored in the DGE_TBNC file. Then, we assign significant labels to the genes if they have P-value \< 0.05. If the log(Fold Change) is above 0.6, the gene is significantly *up* regulated. If the log(FoldChange) is below -0.6, the gene is *down* regulated.

```{r}
Volcano_DGE <- DGE_TNBC$table #Pulling the table from the DGE including the log Fold Change (FC), P-value and gene symbols
Volcano_DGE$diffexpressed <- "NO"

Volcano_DGE$diffexpressed[Volcano_DGE$logFC > 0.6 & Volcano_DGE$PValue < 0.05] <- "UP"

Volcano_DGE$diffexpressed[Volcano_DGE$logFC < -0.6 & Volcano_DGE$PValue < 0.05] <- "DOWN"
#Assigning the differential expression classification if the logFC and P-value have statistically significant results.  
```

Let's begin with a simple plot using ggplot2.

```{r}
#Basic volcano plot
ggplot(Volcano_DGE, aes(x=logFC, y= -log10(PValue))) +
  geom_point(size=0.5)

```

```{r}
ggplot(Volcano_DGE, aes(logFC, -log10(PValue), col = diffexpressed)) + 
  geom_vline(xintercept = c(-0.6, 0.6), col = "grey", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05), col = "grey", linetype = 'dashed') +
  geom_point(size = 0.5) +
  scale_color_manual(values = c("blue", "grey", "red"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
coord_cartesian(ylim = c(0, 50), xlim = c(-10, 10)) + # since some genes can have minuslog10padj of inf, we set these limits
  labs(color = 'Severe', #legend_title, 
       x = expression("log"[2]*"FC"), y = expression("-log"[10]*"p-value")) + 
  scale_x_continuous(breaks = seq(-10, 10, 2)) +  
  theme_minimal() +
  geom_text_repel(
    data = subset(Volcano_DGE, (abs(logFC)) > 1 & PValue < 1e-5),
    aes(label = Symbol),
    size = 3
  )

```

# Gene ontology pathway

Moving onto gene ontology, this is where our now identified differential genes will be grouped into their respective classification. We will be using the DGE_TNBC file we created above.

*Important reminders:* The DGE file created above follows the contrast: Healthy - TNBC cells. All of the positive genes are higher

```{r}
DGE_gene_list <- DGE_TNBC$table  #Suppress warning is to limit the warnings producted during the gseGO command.
  geneList <- DGE_gene_list$logFC
  names(geneList) <- rownames(DGE_gene_list)
  geneList <- sort(geneList, decreasing = TRUE)
  geneList <- na.omit(geneList)

require(org.Hs.eg.db)

gse <- gseGO(geneList=geneList, 
             ont ="BP", #The current setting is the biological purpose. Feel free to change this for MF (molecular function), CC (cellular component) or ALL for all pathways
             keyType = "ENSEMBL", 
             nPerm = 10000, 
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = org.Hs.eg.db, #Organism database = human
             pAdjustMethod = "none")
```

### Dot plot of GO

Below generates a dot plot of the top and bottom 5 categories impacted in the differential genes.

-   p.adjust corresponds to whether the gene is up or down regulated.

-   Count is the number of genes involved in that pathway impacted

```{r}
require(DOSE)

options(enrichplot.colours = c("red","blue"), font.size=6) #picks the colors, feel free to change as you desire. R uses simple color names or hex codes.
```

```{r}
dotplot(gse, showCategory = 5, split =".sign") + 
  facet_grid(.~.sign) + #Facets thegraph by the upregulated (activated) or downregulated (suppressed)
  theme_linedraw() + theme(panel.grid = element_line(color = "lightgrey"))

```

From the plot, the TNBC condition has the following pathways activated:

-   Nuclear chromosome segregation

-   Lymphocyte mediated immunity

-   Chromosome organization

-   Chromosome segregation

-   Nuclear division

### Pairwise gene map

Creates a map connecting the pathways impacted to each other. In TNBC, we see nuclear division is linked to meiosis and mitotic division.

```{r}
gse_sim <- pairwise_termsim(gse)
emapplot(gse_sim, showCategory = 10)
```

